Add HPKE DHKEM provider support for EC, X25519 and X448.

slontis opened the pull request:
This is a implementation of the HPKE DHKEM for EC and ECX.
The KEM algorithm is implemented using the KEM provider interface.
Extra APIS were added to support authentication.
The interface for deriving keys using DHKEM uses the keygen API, by supplying the seed (ikm) as a settable param

<!--
Thank you for your pull request. Please review these requirements:

Contributors guide: https://github.com/openssl/openssl/blob/master/CONTRIBUTING.md

Other than that, provide a description above this comment if there isn't one already

If this fixes a GitHub issue, make sure to have a line saying 'Fixes #XXXX' (without quotes) in the commit message.
-->

##### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->
- [x] documentation is added or updated
- [x] tests are added or updated

paulidale commented:
CIs are relevant.

slontis commented:
@sftcd this is a first pass, that just does the X25519 & X448 KEM.
It includes code that is shared with the rest of hpke in the hpke_util.c
We will have to have a talk about that I guess - but you can modify it to suit your needs.
(In order to commit this I obviously need those bits for it to work).

Rather than jamming an API into the KEM to do the keyderive -  I have added to the keygen which seems to fit much nicer (i,.e. this gets rid of the dummy key I was creating)_

The supplied test shows the interfaces...
I have added an EVP_PKEY_auth_encapsulate_init() and EVP_PKEY_auth_decapsulate_init() for the auth case.
The KEM already had an EVP_PKEY_encapsulate_init()/EVP_PKEY_encapsulate() interface.

Inside the DHKEM code it it setting up all the params related to kem_id, kdf and kdfdigest based on whether it is X25519 or X448. If we were to require overrides in the future they could be set via set_params() - but they are not really needed.
Let me know if you want to set up a meeting at some point..

slontis commented on providers/implementations/keymgmt/ecx_kmgmt.c:
```diff
@@ -594,8 +607,25 @@ static void *ecx_gen(struct ecx_gen_ctx *gctx)
+            goto err;
+        if (!ossl_ecx_public_from_private(key))
+            goto err;
+        key->haspubkey = 1;
```
@mattcaswell moving the code into here shows that it is not doing the masking on the private key buffer as is done further down in this code block.. Is this a problem?

t8m replied:
What is the masking below good for? Is it required for the X*** private keys to work properly?

mattcaswell replied:
See https://www.rfc-editor.org/rfc/rfc7748 section 5:

>    Scalars are assumed to be randomly generated bytes.  For X25519, in
   order to decode 32 random bytes as an integer scalar, set the three
   least significant bits of the first byte and the most significant bit
   of the last to zero, set the second most significant bit of the last
   byte to 1 and, finally, decode as little-endian.  This means that the
   resulting integer is of the form 2^254 plus eight times a value
   between 0 and 2^251 - 1 (inclusive).  Likewise, for X448, set the two
   least significant bits of the first byte to 0, and the most
   significant bit of the last byte to 1.  This means that the resulting
   integer is of the form 2^447 plus four times a value between 0 and
   2^445 - 1 (inclusive).

mattcaswell replied:
So, is the output from `ossl_ecx_dhkem_derive_private` just a random 32 byte string? In which case it will need to be masked. Or has it already been masked?

slontis replied:
The description listed at https://www.rfc-editor.org/rfc/rfc9180.html#name-derivekeypair does not do this..
The test vectors listed in the RFC (e.g https://www.rfc-editor.org/rfc/rfc9180.html#name-base-setup-information) do not have the masking..
e.g.

ikmE: (The seed)
7268600d403fce431561aef583ee1613527cff655c1343f29812e66706df3234
pkEm: (public bytes)
37fda3567bdbd628e88668c3c8d7e97d1d1253b6d4ea6d44c150f741f1bf4431
skEm: (private bytes)
52c4a758a802cd8b936eceea314432798d5baf2d7e9235dc084ab1b9cfa2f736

slontis replied:
There is another section [7.1.2. ](https://www.rfc-editor.org/rfc/rfc9180.html#section-7.1.2)[SerializePrivateKey and DeserializePrivateKey](https://www.rfc-editor.org/rfc/rfc9180.html#name-serializeprivatekey-and-des) that does this by the looks of it..

slontis replied:
Not sure those test vectors are correct.. As it says they are in serialized form,,

mattcaswell replied:
Probably the masking needs to be done for these keys too.

slontis replied:
Added. The tests need to do the masking in order to counter a deficiency in the test vectors.

sftcd commented:
> @sftcd this is a first pass, that just does the X25519 & X448 KEM. It includes code that is shared with the rest of hpke in hpke_util.c We will have to have a talk about that I guess 

Good stuff. I'll ping ya for a time to chat next week.

sftcd commented:
Just started to play with this - added a bit of test code for when doing HPKE encryption, just for x25519 and with base mode, and not pushed anywhere yet. I'm getting a failure at the 2nd call to ``EVP_PKEY_encapsulate()`` if no initial key material (ikm) is supplied (or if a param with a NULL/O ikm is supplied) - what's the intent for that case? RFC9180 says that means to generate some random ikm, but should the HPKE code generate the random ikm and give that to your provider or should generation of the random ikm happen within the provider if no specific ikm is supplied by the caller? My initial guess was the latter but seems that's not working. (For ECH usage, I think the norm will be to not have any caller-supplied ikm from TLS clients.)

slontis commented:
It doesnt do it currently.. I wasnt sure if I should do that internally if no ikm is supplied, or whether the caller does it..

sftcd commented:
> It doesnt do it currently.. I wasnt sure if I should do that internally if no ikm is supplied, or whether the caller does it..

I'd say it might be better to generate the random ikm inside the provider, for two reasons: 1) that code should know better how many random octets to use and whether e.g. some h/w is available and 2) I'd guess some applications might be better off never knowing how to re-generate the same private key. That said, for ECH it doesn't make much difference, so long as the sender's ephemeral private key can be used a 2nd time by the same process in the event of a HelloRetryRequest.

slontis commented:
Ok will change it - its not a big change...

slontis commented:
Do you have any comment on the test vector for X25519 private keys..
Looking at RFC 9180 A.1.1

it has this
skEm:
52c4a758a802cd8b936eceea314432798d5baf2d7e9235dc084ab1b9cfa2f736

The initial section describes this as being in serialized form...
Which should do this to the private key...
        privkey[0] &= 248;
        privkey[X25519_KEYLEN - 1] &= 127;
        privkey[X25519_KEYLEN - 1] |= 64;
        
        When it does this the first byte is then 0x50 and the last byte is 0x76>>
        Am I interpreting this incorrectly?

sftcd commented:
Yeah I think the above is correct (as per https://www.rfc-editor.org/rfc/rfc7748#section-5) , but not sure your new code needs to do that - for test purposes I just compare the result of using the IKM to the generated public key, which should be good enough I think.

slontis commented:
I will do the masking in the test code.. I really think the test vector should be updated to be correct.

slontis commented:
Do you happen to have any X448 KEM test vectors?

sftcd commented:
IIRC there are some in the test-vectors.json file at https://github.com/cfrg/draft-irtf-cfrg-hpke/

sftcd commented:
On the good news front, when I do provide ikm, and whack in a call to your kem stuff into my shim code, my pre-existing decryption code works fine (for base mode, x25519 anyway). Seems like a good sign:-) https://github.com/sftcd/happykey/blob/ecxdhkem/hpke.c#L2160

sftcd commented:
Just to ensure we're on the same page - after I make a call to your kem API when encrypting, My HPKE code has to do a bunch of extracts and expands using the shared secret output from your kem API before I call the aead function. One example label input to the first of those extracts is "psk_id_hash", so that string shouldn't occur in your code but just in mine. I'm just checking that I'm clear on where we want the boundary to be before we start thinking about other curves modes etc. 

Put another way, the code from [here](https://github.com/sftcd/happykey/blob/ecxdhkem/hpke.c#L2244) to the end of that function wouldn't be changed due to this PR. (It'll change based on other comments and I'll be doing the 9180 based API stuff, but ignore that for the moment.) 

Sound about right?

slontis commented:
Yes I am only doing the KEM part..(Encap / Decap)
The key schedule is a separate operation that belongs in  HPKE..
I would suggest putting the schedule into a separate internal function.

slontis commented:
The keygen related changes are just to create the recipient (and sender auth) keys from ikm values..

slontis commented:
I am getting close with the EC code now.. Just need to test and write a bit of code for the auth mode.. I am hoping to get close to submitting by the end of my day today..

sftcd commented:
Grand so. I can play more with it tomorrow (here) and/or Wed. Just send me a ping here or via mail.

slontis commented:
@sftcd - just getting the errors out of the CI loop now..
The interface is very similiar to the ecx case.

sftcd commented:
Just checked that latest works for all groups in base mode and it does, which is neat. Will check with other HPKE modes tomorrow.

Looks like a non-NULL ikm still has to be supplied in this version, and (perhaps worse?) the ikm length has to match the private key length for the group. That's a level of clue it's a bit hard to expect from a caller IMO. I'd argue both restrictions ought go, noting that length of input != entropy of input if someone says you need that many random bits of ikm to be supplied by the caller:-)

slontis commented:
Generation of random IKM in EC has been added.. Documentation is also updated.

sftcd commented:
> Generation of random IKME inside the EC KEM for the sender ephemeral key has been added.. Documentation is also updated.

... and confirmed to work, thanks. (Where "work" so far is still just that code I whacked into my test shim to use this when encrypting and older code when decrypting, but it'll be same code.)

slontis commented:
Fairly comprehensive Coverage testing added..

slontis commented:
ping
This is now ready for review.

slontis commented:
ping
can someone review this please..

sftcd commented:
On 06/09/2022 10:31, Shane wrote:
> ping
> can someone review this please..

I'll be looking at it again next day or two if that helps
(though I guess you're looking from reviews from committers
so it probably doesn't:-)

S.

sftcd commented:
(Starting to try this out more with the new HPKE API stuff now, and thanks for your comments on that, I'll respond to those in a bit...)

Am I right in reading this as follows: to use the HPKE AUTH modes one needs to make calls to e.g. ``EVP_PKEY_auth_encapsulate_init()`` whereas for the base modes one calls ``EVP_PKEY_encapsulate_init()``? 

If so, won't that lead to lots more code that is effectively only called by the test harness? (There being no current uses of HPKE AUTH modes afaik.)

That'd also mean similar increases in lines of code in the HPKE PR I think, compared to APIs that allow setting the AUTH private or public when needed as an attribute of the context fed into one ``EVP_PKEY_encapsulate_init()`` call. 

I'm not seeing the benefit of that approach, what am I missing? Or maybe I'm reading it wrong?

slontis commented:
I cant change the existing API.. It is now fixed in stone.. Hence a new API.

slontis commented:
if (auth)
  auth_encapsulate_initi()
else
  encapsulate_init()
  
Is not a lot of code :)

sftcd commented:
Sorry, not following. What's ``EVP_PKEY_encapsulate_init()`` currently used for? Somewhere in there you must be calling the new ``hpke_expand()`` I guess, so not clear what calls for ``EVP_PKEY_auth_encapsulate_init()``. Maybe I need to read it more:-)

(If it were just one call to _init() that'd be easy, sure but there's a bunch of 'em needed)

sftcd commented:
ah sorry, reading more it's less bad than I figured - so the same EVP_PKEY_CTX var can be used in both cases? If so, that;'s ok

sftcd commented:
ah sorry, reading more it's less bad than I figured - so the same EVP_PKEY_CTX var can be used in both cases? If so, that;'s ok

slontis commented:
The pattern is this

EVP_PKEY_encapsulate_init(key)
//Potentially do other things such as setparams()
EVP_PKEY_encapsulate(enc, secret)

This init(), set(), dooperation() is a common pattern for most operations in openssl.

slontis commented:
Yes, to create a sender you create a recipient key and use that for the ctx.. that is then used by either init().
My tests just then go  if (data->authkey) do the auth_encapsulate_init() else do the encapsulate_init() followed by the encapsulate() call that is used by both.. My thinking is that maybe the auth mode does something completely different for some new kem that comes along at some point...

slontis commented:
The one remaining thing I havent done (which can be done later) is limit the ikm length. Currently it will just fail in the extract or expand functions since they use fixed size buffers. For interop reasons I think you should be applying the 64 byte limits to the other HPKE fields that are listed in the RFC.. The code may need to do 2 passes when using the wpacket code, one to calculate the size (and if the size is too large for the fixed buffer - then use a malloced buffer instead).

sftcd commented:
making progress with the AUTH mode stuff now, not sure I'll reach success tonight (late here), but direction == good:-)

slontis commented:
> making progress with the AUTH mode stuff now, not sure I'll reach success tonight (late here), but direction == good:-)

Excellent. You have some ref code in my old PR that worked with auth 😉. The DHKEM stuff should be good as it has been fairly well tested.

slontis commented:
rebased to fix test/build.info conflict

sftcd commented:
I've confirmed this all works for my [(shim) HPKE build](https://github.com/sftcd/happykey/tree/ecxdhkem2) so as soon as this is merged I can simply my HPKE PR #17172 significantly and easily ensure this works for ECH. IOW, please merge soon:-)

mattcaswell commented on providers/implementations/kem/ec_kem.c:
```diff
@@ -0,0 +1,842 @@
+/*
+ * Copyright 2022 The OpenSSL Project Authors. All Rights Reserved.
```
Shouldn't the name of this source file reference dhkem in some way, e.g. ec_dhkem.c

slontis replied:
No it shouldnt...
There is a ctx->mode which is set to KEM_MODE_DHKEM..
There could possible be other EC KEM types in future..
This decision was made when we did the RSA KEM... i.e. You have to set OSSL_KEM_PARAM_OPERATION before using it (there is no default).

mattcaswell commented on providers/implementations/kem/ec_kem.c:
```diff
@@ -0,0 +1,842 @@
+    size_t ikmlen;
+    const char *kdfname;
+    const DHKEM_ALG *alg;
+
```
Remove the blank line

mattcaswell commented on providers/implementations/kem/ec_kem.c:
```diff
@@ -0,0 +1,842 @@
+static OSSL_FUNC_kem_set_ctx_params_fn eckem_set_ctx_params;
+static OSSL_FUNC_kem_settable_ctx_params_fn eckem_settable_ctx_params;
+
+/* Move this table if it is needed by the rest of HPKE */
```
This comment is transient and only relevant during the implementation of HPKE. I'd remove this before merge.

slontis replied:
Removed and replaced with ref to Table 7.1

mattcaswell commented on providers/implementations/kem/ec_kem.c:
```diff
@@ -0,0 +1,842 @@
+    if (kdfctx == NULL)
+        return 0;
+
+
```
Too many blank lines

mattcaswell commented on providers/implementations/kem/ec_kem.c:
```diff
@@ -0,0 +1,842 @@
+    const EC_GROUP *group = EC_KEY_get0_group(sender);
+    size_t secretlen = (EC_GROUP_get_degree(group) + 7) / 8;
+
+
```
Too many blank lines

mattcaswell commented on providers/implementations/kem/ecx_kem.c:
```diff
@@ -0,0 +1,704 @@
+    const char *kdfdigestname;
+    size_t sharedsecretlen;
+    size_t keylen;
+
```
Remove the blank line

mattcaswell commented on providers/implementations/kem/ecx_kem.c:
```diff
@@ -0,0 +1,704 @@
+        if (!ossl_ecx_key_up_ref(ecx))
+            return 0;
+        ctx->recipient_key = ecx;
+
```
Remove the blank line

mattcaswell commented on providers/implementations/kem/ecx_kem.c:
```diff
@@ -0,0 +1,704 @@
+                                   const unsigned char *pubbuf, size_t pubbuflen)
+{
+    ECX_KEY *ecx = NULL;
+    OSSL_PARAM params[3], *p = params;
```
Why 3 and not 2?

mattcaswell commented on providers/implementations/kem/ecx_kem.c:
```diff
@@ -0,0 +1,704 @@
+    }
+
+    if (ecx->type == ECX_KEY_TYPE_X25519) {
+        kemid = 0x20;
```
What is this magic number?

slontis replied:
Shifted to a #define with a description, plus made this code into a function that is shared by the 2 places that use it.

mattcaswell commented on test/evp_pkey_dhkem_test.c:
```diff
@@ -0,0 +1,871 @@
+
+#include "testutil.h"
+
+#ifndef OPENSSL_NO_EC
```
We basically seem to skip everything if EC is not enabled. Why not avoid building this file in build.info if there is no ec. Similarly skip the whole thing in 30-test_evp_pkey_dhkem.t if no-ec. That way we can avoid all the ifdefs in this file.

slontis replied:
ok

mattcaswell commented on test/evp_pkey_dhkem_test.c:
```diff
@@ -0,0 +1,871 @@
+        /*
+         * The RFC test vector seems incorrect since it is not in serialized form,
+         * So manually do the conversion here for now.
+         */
```
Have we raised an errata entry for this against the RFC?

slontis replied:
Yes I have had a discussion with the authors and raised an errata. I assume other implementations must just be testing the public key (since it is derived from the private).

mattcaswell reviewed (changes requested):
We should probably discuss the proposed API changes at OTC (or did we already do so when we discussed HPKE previously??).

slontis commented:
> We should probably discuss the proposed API changes at OTC (or did we already do so when we discussed HPKE previously??).

I redid quite a few things since the initial discussion, so I am happy to discuss again..

sftcd commented:
On 12/09/2022 22:32, Shane wrote:
>> We should probably discuss the proposed API changes at OTC (or did
>> we already do so when we discussed HPKE previously??).
> 
> I redid quite a few things since the initial discussion, so I am
> happy to discuss again..

I don't think I'm needed for that discussion but would be
happy to take part if that helped.

I can say Shane's current API works fine for my HPKE PR, for
all combinations of HPKE mode, and suite (kem/kdf/aead) when
round-tripping. It also works fine for all the test vectors
I've tried so far. It was also easy enough to figure out how
to use it from the test code, which seems pretty thorough.

It'll also help me a lot to see this merged; before then it's
a bit tricky to test the combination of Shane's KEM stuff,
and my HPKE stuff and ECH stuff all together and it's that
combo that's the important interop target for the near term,
i.e. OpenSSL-enabled clients working with CF's servers and
browsers working with servers using OpenSSL. If it were going
to be a while before this were merged (which'd be a pity), it
would be good to know that so I can figure a way to do those
interop tests on an ongoing basis.

Lastly, for now, I also plan to maintain a version of my HPKE
shim code that works against a released library. That should
be potentially useful until there's a library release that
has HPKE, which I guess will be some time (but if I'm wrong
there, be good to know that).

Cheers,
S.

sftcd commented:
On 12/09/2022 22:41, Shane wrote:
> Yes I have had a discussion with the authors and raised an errata. I
> assume other implementations must just be testing the public key
> (since it is derived from the private).
One can argue that that's not wrong. IIRC implementations
presented with random octet strings are supposed to do the
masking anyway so from that POV it doesn't matter that the
output from the deterministic key gen isn't masked and is
used for test vectors, but on the whole it'd probably have
been better if the key gen had said to do the masking and
the test vectors had been masked. But, it could be the
right outcome is to just document that the test vectors in
the RFC aren't masked.

Cheers,
S.

slontis commented on crypto/evp/kem.c:
```diff
@@ -159,6 +167,24 @@ static int evp_kem_init(EVP_PKEY_CTX *ctx, int operation,
         }
         ret = kem->decapsulate_init(ctx->op.encap.algctx, provkey, params);
         break;
+    case EVP_PKEY_OP_AUTH_ENCAPSULATE:
```
I was thinking that maybe the init may need to be completely different in some future case. I originally had it as a setter. You still end up with another function to do the set in order to pass the ECX  or EC key in.

slontis commented on crypto/ec/ec_key.c:
```diff
@@ -350,6 +351,76 @@ static int ec_generate_key(EC_KEY *eckey, int pairwise_test)
+            EC_POINT_set_to_infinity(group, eckey->pub_key);
+    }
+    return ok;
+}
```
ossl_ec_key_simple_generate_public_key() is basically the point_multiply line..
Not sure that warrants melding the two? I could call this code from the generate but that doesnt make things simpler.

mattcaswell replied:
I guess my concern was why does that `keygenpub` function pointer exist in EC_METHOD at all? Are there circumstances where you would need a different implementation? I can't think what those circumstances might be, and AFAICT there don't seem to be any alternative implementations anyway. If anyone *does* introduce such an implementation then this could introduce a subtle bug here because it wouldn't be used.

It also seems odd to have that function, and then introduce a second essentially identical implementation.

slontis replied:
I have got rid of the function as it was only needed internally.. Removing the function does to some extent mean the code I needed just shifted somewhere else, so this doesnt achieve very much. I was trying to make the code look like the other generate function. I call the function used by keygenpub to run the couple of lines of code to do a multiply now. Note that the generate() function doesnt do that, which is what I was trying to copy. NOTE: The keygenpub() is called in one place by a d2i method.

slontis commented:
Addressed review comments..

slontis commented:
rebased with no changes..

levitte commented on doc/man7/OSSL_PROVIDER-default.pod:
```diff
@@ -193,6 +193,8 @@ The OpenSSL default provider supports these operations and algorithms:
 
 =item RSA, see L<EVP_KEM-RSA(7)>
+=item X25519, see L<EVP_KEM-X25519(7)>
+=item EC, see L<EVP_KEM-EC(7)>
```
You *must* have empty lines separating the `=item` lines.  I know it's annoying, but the POD reader strictly reads in paragraph mode, so without the empty lines, what you get here is one item, looking like this:

```
=item RSA, see L<EVP_KEM-RSA(7)> =item X25519, see L<EVP_KEM-X25519(7)> =item EC, see L<EVP_KEM-EC(7)>
```

slontis replied:
Thanks, I forget that sometimes..

levitte commented on include/openssl/evp.h:
```diff
@@ -1696,6 +1696,8 @@ const char *EVP_PKEY_get0_type_name(const EVP_PKEY *key);
 # define EVP_PKEY_OP_ENCAPSULATE         (1<<12)
 # define EVP_PKEY_OP_DECAPSULATE         (1<<13)
+# define EVP_PKEY_OP_AUTH_ENCAPSULATE    (1<<14)
+# define EVP_PKEY_OP_AUTH_DECAPSULATE    (1<<15)
```
Drop these.  Instead, there should be checks in libcrypto code along the lines of this:

``` C
        if (provauthkey == NULL && kem->decapsulate_init != NULL) {
            ret = kem->decapsulate_init(ctx->op.encap.algctx, provkey, params);
        } else if (provauthkey != NULL && kem->auth_encapsulate_init != NULL) {
            ret = kem->auth_encapsulate_init(ctx->op.encap.algctx, provkey,
                                             provauthkey, params);
       } else {
            ERR_raise(ERR_LIB_EVP, EVP_R_OPERATION_NOT_SUPPORTED_FOR_THIS_KEYTYPE);
            ret = -2;
            goto err;
        }
```

slontis replied:
Thanks @levitte

slontis commented:
removed extra #defines for auth encapsulate/decapsulate ops
@mattcaswell I think the only outstanding thing now is the question related to the public from private function..

slontis commented:
rebased to fix merge conflicts.
Changes: Remove extra spaces at end of lines in pod files
                Collapsed all commits into 1 commit + the merge conflict commit.
                Changed the commit message, to a summary of all commits.

slontis commented:
ping - Can someone review this please

slontis commented:
Buitblot failures are not related to this PR.

t8m commented:
The commit messages do not look correct. Could you please amend them so the commit title is meaningful, etc.

slontis commented:
Commit message updated..

t8m commented on providers/defltprov.c:
```diff
@@ -412,6 +412,11 @@ static const OSSL_ALGORITHM deflt_asym_cipher[] = {
+#ifndef OPENSSL_NO_EC
+    { PROV_NAMES_X25519, "provider=default", ossl_ecx_asym_kem_functions },
+    { PROV_NAMES_X448, "provider=default", ossl_ecx_asym_kem_functions },
+    { PROV_NAMES_EC, "provider=default", ossl_ec_asym_kem_functions },
```
> Another possible naming approach could be to have a single algorithm called "DHKEM" and extend the "groups" that can be passed to include X25519 and X448 and well as normal EC groups.

That would be IMO very problematic from the implementation standpoint because the key types for these groups are different.

mattcaswell replied:
> That would be IMO very problematic from the implementation standpoint because the key types for these groups are different.

Isn't that just a matter of special casing the X25519/X448 "groups". Although I suppose X25519 isn't really a group at all (Curve25519 is)...so maybe that doesn't work

t8m replied:
Thinking about it more - the current way how libcrypto implements the evp_kem_init() does not allow for specifying a different type of KEM on a given key type - what evp_keymgmt_util_query_operation_name() returns is used. Unless we really plan to extend libcrypto API to allow that, there is no point in changing the name of the KEM algorithm to confuse things, or we should have already done it with RSA.

mattcaswell replied:
Hmmm. That seems like a fairly nasty limitation of our current API. We really should have some mechanism of determining which KEM algorithm we want to apply. Not least because if a provider decided to implement some other KEM for (say) EC, then suddenly HPKE is not going to work anymore. So, effectively we are locking in "DHKEM" as *the* KEM for use with EC/X25519/X448 forever.

But, ok, that's probably not a problem to be solved in *this* PR.

t8m replied:
> Isn't that just a matter of special casing the X25519/X448 "groups".

The point is that the keymgmt key object is passed to the operation in the provider. There is no way for the encapsulate_init to recognize what type is the key object unless the algorithm fetched is different.

slontis replied:
EVP_PKEY_CTX_set_kem_op() is the way we currently set the operation type.. If the init needs to be really different it can be defered to after the op is set. (This is a settable param so it can also be passed to the init).

I think before we go fixing 'groups' here we may want to consider if having a single EC keymanagement object to represent all curve types is a good idea or not.. I also think this is out of scope within this PR and needs a bigger discussion within OTC.
I initially had shared code to handle both EC and ECX and it was horrible and unreadable. Separating them out simplified the implementations (even if it is similiar looking code for each - it still has to handle the keys differently).

t8m replied:
> EVP_PKEY_CTX_set_kem_op() is the way we currently set the operation type.. If the init needs to be really different it can be defered to after the op is set. (This is a settable param so it can also be passed to the init).

Heh, interesting, I did not know about this function. So we basically have the algorithm choice as a parameter that can be freely ignored by the provider. IMO not too nice. Unfortunately we have to live with that.

IMO merging EC and ECX key types is not a good idea, and definitely not something that should be considered earlier than in 4.0.

slontis commented:
Issues have been addressed..

mattcaswell commented on crypto/ec/ec_key.c:
```diff
@@ -350,6 +351,43 @@ static int ec_generate_key(EC_KEY *eckey, int pairwise_test)
+        if (eckey->pub_key == NULL)
+            goto err;
+    }
+    if (!ossl_ec_key_simple_generate_public_key(eckey))
```
Probably this should be `eckey->group->meth->keygenpub(eckey)`.

slontis replied:
I  can now see where you are going with this, however this code is a variation of ec_generate_key().
If this is wrong then so is ossl_ec_key_pairwise_check(). So this to me looks more like a refactor task that is out of scope for this PR.

slontis replied:
@mattcaswell any additional input here?

slontis replied:
If it had of just fit into the existing keygen as an extra 'if else clause' there would probably not have been a question related to this :)

slontis commented:
ping - Can I get a second reviewer please..

hlandau commented on providers/implementations/kem/ec_kem.c:
```diff
@@ -0,0 +1,839 @@
+static OSSL_FUNC_kem_set_ctx_params_fn eckem_set_ctx_params;
+static OSSL_FUNC_kem_settable_ctx_params_fn eckem_settable_ctx_params;
+
+/* See Table 7.1 */
```
Should be referenced as Section 7.1 or Table 2

hlandau commented on providers/implementations/kem/ec_kem.c:
```diff
@@ -0,0 +1,839 @@
+static OSSL_FUNC_kem_settable_ctx_params_fn eckem_settable_ctx_params;
+
+/* See Table 7.1 */
+static const DHKEM_ALG dhkem_alg[]= {
```
Space before =

hlandau commented on providers/implementations/kem/ec_kem.c:
```diff
@@ -0,0 +1,839 @@
+
+/*
+ * Used to store constants from Table 7.1
+ * and the bitmask for curves described in Section 7.1.3 DeriveKeyPair
```
Should be referenced as Section 7.1 or Table 2

hlandau commented on test/dhkem_test.inc:
```diff
@@ -0,0 +1,707 @@
+        p256_expected_enc, sizeof(p256_expected_enc),
+        p256_expected_secret, sizeof(p256_expected_secret),
+    },
+    {
```
I would suggest adding the auth test vector for P256 from the RFC also.

slontis replied:
I left it out since the HPKE tests are likely to rehit all of theses cases... I just did p521 as it is the largest data set, so the combos of p-256 non auth, p-521 non auth and p521 auth should be hitting all the bits of code that actually matter..

hlandau reviewed (changes requested):
Very nice. Just some small nits, once fixed I'll approve.

slontis commented on crypto/hpke/hpke_util.c:
```diff
@@ -0,0 +1,174 @@
+ */
+#define LABELED_EXPAND_SIZE (LABELED_EXTRACT_SIZE + 3 * OSSL_HPKE_MAX_PUBLIC)
+
+static const char LABEL_HPKEV1[] = "HPKE-v1";
```
Changed

slontis replied:
Had to add a few more since there are labels for extract and expand ops inside the ec and ecx kems

slontis commented:
NITS addressed

hlandau commented:
@mattcaswell Still OK?

hlandau commented:
@slontis This needs rebasing now. So close...

hlandau commented:
Merging.

hlandau commented:
Merged to master. Thank you.

slontis commented:
Thanks for reviewing!

@sftcd time to rebase :)

sftcd commented:
On 23/09/2022 09:51, Shane wrote:
> Thanks for reviewing!
> 
> @sftcd time to rebase :)

Will do. Good to see progress progressing.

S.
